### PR TITLE
Add test_list command, to enumerate unit and e2e tests.

### PR DESCRIPTION
**Pull request by rmmh**
This uses go/parser and go/ast to analyze all test files in ~1 second.
It only recognizes a few simple structures that the tests all have, and
modifies a few tests to fit expected structure better.

This is part of an effort to ensure all tests have owners, by having a
verify check to catch new tests being added without an owner.

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/33781)

### CODE REVIEW

**Comment by rmmh** [comment]
Sample of the output (there's a JSON mode for tools, but this is more readable):
```
{cmd/genutils/genutils_test.go:23:1 k8s.io/kubernetes/cmd/genutils TestValidDir}
{cmd/genutils/genutils_test.go:30:1 k8s.io/kubernetes/cmd/genutils TestInvalidDir}
{cmd/genutils/genutils_test.go:37:1 k8s.io/kubernetes/cmd/genutils TestNotDir}
{cmd/hyperkube/hyperkube_test.go:131:1 k8s.io/kubernetes/cmd/hyperkube TestRun}
{cmd/hyperkube/hyperkube_test.go:137:1 k8s.io/kubernetes/cmd/hyperkube TestLinkRun}
...
{test/e2e/service.go:336:2 [k8s.io] Services should work after restarting kube-proxy [Disruptive]}
{test/e2e/service.go:388:2 [k8s.io] Services should work after restarting apiserver [Disruptive]}
{test/e2e/service.go:434:2 [k8s.io] Services should be able to create a functioning NodePort service}
{test/e2e/service.go:465:2 [k8s.io] Services should be able to change the type and ports of a service [Slow]}
{test/e2e/service.go:784:2 [k8s.io] Services should use same NodePort with same port but different protocols}
{test/e2e/service.go:831:2 [k8s.io] Services should prevent NodePort collisions}
```

/cc @kubernetes/sig-testing

**Comment by spxtr** [comment]
@k8s-bot test this https://github.com/kubernetes/test-infra/issues/709

**Comment by apelisse** [comment]
I have almost no knowledge of that, so I read the code. It seems OK to me but it's not worth much :-)

File: test/list/main.go [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+// and staging directories.
+func (t *testList) handlePath(path string, info os.FileInfo, err error) error {
+	if err != nil {
+		return err
+	}
+	if strings.Contains(path, "third_party") ||
+		strings.Contains(path, "staging") {
+		return filepath.SkipDir
+	}
+	if strings.HasSuffix(path, ".go") && strings.Contains(path, "e2e") ||
```
Comment (spiffxp) [inline test/list/main.go]: do we care about differentiating `e2e` vs `node_e2e`?

Comment (rmmh) [inline test/list/main.go]: No, they both use ginkgo in essentially the same way.

**Comment by rmmh** [comment]
@k8s-bot test this kubernetes/test-infra#709

**Comment by krzyzacy** [comment]
Review status: 0 of 6 files reviewed at latest revision, 3 unresolved discussions, some commit checks broke.

---

*[test/e2e/density.go, line 459 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33781#-KSrZw9fS7LdDZxBCwFM:-KSrZw9fS7LdDZxBCwFN:bq1tnen) ([raw file](https://github.com/kubernetes/kubernetes/blob/373653c4428c141ef54e55138d51dc8724cae0b7/test/e2e/density.go#L459)):*
> ```Go
> 		}
> 
> 		name := fmt.Sprintf("[Feature:%s] should allow starting %d pods per node", feature, testArg.podsPerNode)
> ```

er, github don't have the 80 char per line limit?

---

*[test/list/main.go, line 251 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33781#-KSsX6_9OAsu9q78woKZ:-KSsX6_AOxYNd2s-c5dQ:b-fi9hh4) ([raw file](https://github.com/kubernetes/kubernetes/blob/373653c4428c141ef54e55138d51dc8724cae0b7/test/list/main.go#L251)):*
> ```Go
> }
> 
> func main() {
> ```

Document the usage? What's expected from the args?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33781)*

**Comment by krzyzacy** [comment]
Still learning it, but looks fancy!

**Comment by krzyzacy** [comment]
And make sure the tests are all pass :-)

**Comment by rmmh** [comment]
@k8s-bot test this issue: #ignore

**Comment by krzyzacy** [comment]
@k8s-bot LGTM THIS

**Pull request merged**